File: Power-of-Two/poweroftwo.cpp
#include <cstdlib>   // exit macros
#include <iostream>  // I/O stuff
#include <exception> // exception handling
#include <string>    // stol

// Determines if @num is a power of two
//
// Determines if @num is a power of two by taking
// advantage of the fact that any integer that is a power
// of two only has one bit flipped. 
// Hence, if (@num & @num - 1) == 0
// then @num is a power of two.
//
// param num : any integer
// return    : true if @num is a power of two, false if not
bool isPowerOfTwo(int64_t const & num) {
    return ((num & (num - 1)) == 0);
}

// Prints a helpful message :)
void printHelp() {
    std::cout << "Try running this program with an integer command "
              << "line input.\n"
              << "For example:\n\n"
              << "./thisprogram 2" << std::endl << std::endl << std::endl;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        std::cout << "No command line arguments provided.\n";
        printHelp();
        return EXIT_FAILURE;
    }
    
    for (unsigned int i = 1; i < argc; ++i) {
        try {
            int64_t num = std::stol(argv[i]);

            if (isPowerOfTwo(num) && num != 0) {
                std::cout << num << " is a power of two\n";
            } else {
                std::cout << num << " is not a power of two\n";
            }
        }
        catch (std::exception) { /* do nothing */ }
    }

    return EXIT_SUCCESS;
}

File: Power-of-Two/poweroftwo.hs
module PowerOfTwo where
import Data.Bits

-- Determines if @n is a power of two
--
-- Determines if @n is a power of two by
-- taking advantage of the fact that the
-- log base 2 of any power of two will
-- result in a whole number.
-- Note that this method is inaccurate beyond 2^37
--
-- param n : any number
-- return  : true if @n is a power of two, false if not
poweroftwoLog :: Integer -> Bool
poweroftwoLog n = 
    fromIntegral (floor (logBase 2 (fromInteger n)) :: Integer) == (logBase 2 (fromInteger n))

-- Determines if @n is a power of two
--
-- Determines if @n is a power of two by
-- taking advantage of the fact that
-- powers of two have only one bit flipped.
-- Hence, if the result of a bitwise 'and'
-- operation on @n and @n-1 results in 0,
-- then @n is a power of two.
--
-- param n : any whole number
-- return  : true if @n is a power of two, false if not
poweroftwo :: Integer -> Bool
poweroftwo n = ((n - 1) .&. n) == 0

